pkg/controller/controller.go

```diff
@@ -127,30 +128,47 @@ func (haproxy *HAProxyController) BackendDefaults() defaults.Backend {
 
 // OnUpdate regenerate the configuration file of the backend
 func (haproxy *HAProxyController) OnUpdate(cfg ingress.Configuration) ([]byte, error) {
-	data, err := haproxy.template.execute(newControllerConfig(&cfg, haproxy))
+	updatedConfig := newControllerConfig(&cfg, haproxy)
+
+	reconfigureBackends(haproxy, updatedConfig)
+
+	data, err := haproxy.template.execute(updatedConfig)
```

Comment:
There is a benefit to generating the configuration file even if `reloadRequired` is false - having the configuration in sync with in-memory configuration makes it easy to check what is currently active by just inspecting haproxy.cfg - so I would advise to keep this.